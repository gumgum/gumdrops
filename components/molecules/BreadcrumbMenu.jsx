import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

class BreadcrumbMenu extends Component {
    static displayName = 'BreadcrumbMenu';

    static propTypes = {
        path: PropTypes.string.isRequired,
        menu: PropTypes.arrayOf(
            PropTypes.shape({
                title: PropTypes.string,
                path: PropTypes.string.isRequired
            })
        ).isRequired,
        linkComponent: PropTypes.oneOfType([PropTypes.element, PropTypes.func]).isRequired
    };

    state = {
        showMenu: false
    };

    _toggleMenu = () => this.setState(({ showMenu }) => ({ showMenu: !showMenu }));

    render() {
        const { showMenu } = this.state;
        const { path: parentPath, linkComponent: LinkComponent, menu } = this.props;
        const menuClass = cx('gds-bubble__menu', 'gds-bubble__menu--left', 'gds-bubble__menu--sm', {
            'gds-bubble__menu--menu-open': showMenu
        });

        return (
            <div onClick={this._toggleMenu} className="gds-page-header__breadcrumbs-menu">
                <div className="gds-page-header__breadcrumbs-menu-dots" />
                <div className={menuClass}>
                    <ul className="gds-bubble__menu-list">
                        {menu.map(({ title, path }) => (
                            <li key={path} className="gds-bubble__menu-list-item -ellipsis">
                                <LinkComponent
                                    className="gds-bubble__menu-list-link -text-tr-cap"
                                    to={`${parentPath}/${path}`}>
                                    {title || path}
                                </LinkComponent>
                            </li>
                        ))}
                    </ul>
                </div>{' '}
            </div>
        );
    }
}

export default BreadcrumbMenu;
