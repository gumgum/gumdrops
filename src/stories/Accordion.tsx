import React from 'react';
import cx from 'classnames';

interface AccordionCommonProps {
    color?: 'dark' | 'white';
    size?: 'sm' | '';
}

type OmitColor<T> = Omit<T, 'color'>;

export interface AccordionItemProps
    extends AccordionCommonProps,
        OmitColor<React.LiHTMLAttributes<HTMLLIElement>> {
    isOpen?: boolean;
    isLocked?: boolean;
    label: React.ReactChild;
}

export interface AccordionProps
    extends AccordionCommonProps,
        OmitColor<React.HTMLAttributes<Element>> {
    initialAllOpen?: boolean;
}

export const Accordion: React.FC<AccordionProps> = ({
    children,
    color,
    size = '',
    className,
    initialAllOpen,
    ...otherProps
}) => {
    const rootClass = cx('gds-accordion', className, {
        [`gds-accordion--${color}`]: color
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size,
            isOpen: child.props.isOpen ? child.props.isOpen : initialAllOpen
        });
    });

    return (
        <div className={rootClass} {...otherProps}>
            <ul className="gds-accordion-list">{newChildren}</ul>
        </div>
    );
};
