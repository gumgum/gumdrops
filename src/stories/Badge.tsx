import React, { ReactChild, ReactChildren } from 'react';
import cx from 'classnames';
import { Colors } from 'types';

export interface BadgeProps extends React.HTMLAttributes<Element> {
    text?: JSX.Element[] | ReactChildren | ReactChild;
    color?: Colors;
    empty?: boolean;
}

const baseClass = 'gds-badge';

export const Badge: React.FC<BadgeProps> = ({
    color = 'default',
    text,
    empty,
    className,
    ...otherProps
}) => {
    const rootClass = cx(baseClass, className, {
        [`${baseClass}--${color}`]: color,
        [`${baseClass}--empty`]: empty
    });

    return (
        <span className={rootClass} {...otherProps}>
            {!empty && text}
        </span>
    );
};
