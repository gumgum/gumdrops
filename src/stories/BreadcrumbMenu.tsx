import React, { useState } from 'react';
import cx from 'classnames';
import { BreadCrumb, LinkComponent } from './Breadcrumb';

export interface BreadcrumbMenuProps extends React.HTMLAttributes<Element> {
    path: string;
    sectionTitle?: string;
    menu: BreadCrumb[];
    linkComponent: LinkComponent;
}

const buttonStyle = { border: 'none', outline: 'none', padding: 0 };

export const BreadcrumbMenu: React.FC<BreadcrumbMenuProps> = ({
    path: parentPath,
    linkComponent: LinkComponent,
    sectionTitle = 'section',
    menu,
    ...otherProps
}) => {
    const [showMenu, setShowMenu] = useState(false);

    const toggleMenu = (): void => setShowMenu(currentValue => !currentValue);

    const menuClass = cx('gds-bubble__menu', 'gds-bubble__menu--left', 'gds-bubble__menu--sm', {
        'gds-bubble__menu--menu-open': showMenu
    });

    return (
        <div
            onClick={toggleMenu}
            className="gds-page-header__breadcrumbs-menu"
            aria-haspopup="listbox"
            aria-expanded={showMenu}
            {...otherProps}>
            <button
                type="button"
                className="gds-page-header__breadcrumbs-menu-dots"
                style={buttonStyle}
                aria-label={`${showMenu ? 'Close' : 'Open'} ${sectionTitle} menu`}
            />
            {showMenu && (
                <div className={menuClass}>
                    <ul className="gds-bubble__menu-list" aria-label="Section Menu">
                        {menu.map(({ title, path }) => (
                            <li key={path} className="gds-bubble__menu-list-item -ellipsis">
                                <LinkComponent
                                    className="gds-bubble__menu-list-link -text-tr-cap"
                                    to={`${parentPath}/${path}`}>
                                    {title || path}
                                </LinkComponent>
                            </li>
                        ))}
                    </ul>
                </div>
            )}
        </div>
    );
};
