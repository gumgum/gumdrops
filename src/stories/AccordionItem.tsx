import React, { MouseEvent, KeyboardEvent, useState } from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';
import { keyCodes } from 'types';

type ToggleEvent = MouseEvent<HTMLLIElement> | KeyboardEvent<HTMLLIElement>;

export const AccordionItem: React.FC<AccordionItemProps> = ({
    isOpen: initialIsOpen,
    isLocked,
    label,
    size,
    color,
    className,
    children
}) => {
    const [isOpen, setIsOpen] = useState(initialIsOpen);

    const toggleOpen = (event: ToggleEvent): void => {
        event.stopPropagation(); // don't want nested accordions to propagate events
        event.preventDefault(); // prevents page scroll from space key
        const { type } = event;
        const { key } = event as KeyboardEvent;

        if (isLocked) return;
        const allowToggle =
            (type === 'keypress' && [keyCodes.space, keyCodes.enter].includes(key as keyCodes)) ||
            type === 'click';

        if (!allowToggle) return;

        setIsOpen(isOpen => !isOpen);
    };

    const rootClass = cx('gds-accordion__item', className, {
        [`gds-accordion__item--${color}`]: color,
        'gds-accordion__item--active': isOpen
    });

    const titleClass = cx('gds-accordion__item-title', {
        [`gds-accordion__item-title--${size}`]: size
    });

    const iconClass = cx('gds-accordion__item-icon', {
        [`gds-accordion__item-icon--${size}`]: size
    });

    const childClass = cx('gds-accordion__child-items', {
        [`gds-accordion__child-items--${size}`]: size
    });

    const newChildren = React.Children.map(children, child => {
        if (!React.isValidElement<AccordionItemProps>(child)) {
            return child;
        }

        return React.cloneElement<AccordionItemProps>(child, {
            color,
            size
        });
    });

    const regionId = `AccordionItem-region-${label}`;
    const labelId = `AccordionItem-label-${label}`;

    return (
        <li
            aria-pressed={isOpen}
            aria-controls={regionId}
            aria-expanded={isOpen}
            className={rootClass}
            onClick={toggleOpen}
            onKeyPress={toggleOpen}
            role="button"
            tabIndex={isLocked ? -1 : 0}>
            <h4 id={labelId} className={titleClass}>
                {label}
            </h4>
            {!isLocked && <span className={iconClass} />}
            <ul
                aria-labelledby={labelId}
                aria-hidden={!isOpen}
                className={childClass}
                id={regionId}
                role="region">
                {newChildren}
            </ul>
        </li>
    );
};
