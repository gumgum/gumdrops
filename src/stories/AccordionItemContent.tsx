import React from 'react';
import cx from 'classnames';
import { AccordionItemProps } from './Accordion';

export const AccordionItemContent: React.FC<Omit<AccordionItemProps, 'label'>> = ({
    className,
    color,
    size,
    children,
    ...otherProps
}) => {
    const stopClick: React.MouseEventHandler = e => {
        e.stopPropagation();
    };

    const rootClass = cx('gds-accordion__child-item', className, {
        [`gds-accordion__child-item--${color}`]: color
    });

    const titleClass = cx('gds-accordion__child-item-title', {
        [`gds-accordion__child-item-title--${size}`]: size
    });

    return (
        <li className={rootClass} onClick={stopClick} {...otherProps}>
            <h4 className={titleClass}>{children}</h4>
        </li>
    );
};
