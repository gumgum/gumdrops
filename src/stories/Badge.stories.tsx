import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Badge, BadgeProps } from './Badge';
import { Colors } from '../types';

export default {
    title: 'Badge',
    component: Badge
} as Meta;

const Template: Story<BadgeProps> = args => <Badge {...args} />;

export const Primary = Template.bind({});
Primary.args = {
    color: Colors.primary,
    text: 'Primary badge'
};

export const Warning = Template.bind({});
Warning.args = {
    color: Colors.warning,
    text: 'Warning Badge'
};

export const Empty = Template.bind({});
Empty.args = {
    color: Colors.info,
    empty: true
};
