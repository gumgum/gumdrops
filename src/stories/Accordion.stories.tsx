import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import { Accordion, AccordionProps } from './Accordion';
import { AccordionItem } from './AccordionItem';
import { AccordionItemContent } from './AccordionItemContent';

export default {
    title: 'Accordion',
    component: Accordion
} as Meta;

const Template: Story<AccordionProps> = args => (
    <Accordion {...args}>
        <AccordionItem label="I'm locked and open" isLocked isOpen>
            <AccordionItemContent>Locked Content</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 1">
            <AccordionItemContent>Content 1</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 2">
            <AccordionItemContent>Content 2</AccordionItemContent>
        </AccordionItem>
        <AccordionItem label="Item 3">
            <AccordionItemContent>Content 3</AccordionItemContent>
            <Accordion>
                <AccordionItem label="Nested Item">
                    <AccordionItemContent>Nested Content</AccordionItemContent>
                </AccordionItem>
            </Accordion>
        </AccordionItem>
    </Accordion>
);

export const Primary = Template.bind({});
Primary.args = {};
