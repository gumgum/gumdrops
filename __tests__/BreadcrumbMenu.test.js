import React from 'react';
import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import { BreadcrumbMenu } from 'stories/BreadcrumbMenu';

describe('BreadcrumbMenu', () => {
    it('renders without option button', () => {
        const menu = [{ title: 'Bar', path: 'bar' }, { path: 'baz' }];
        const TestComponent = props => <div {...props}></div>;
        const { container } = render(
            <BreadcrumbMenu path="/foo" menu={menu} linkComponent={TestComponent} />
        );
        expect(screen.queryAllByRole('listitem').length).toBe(0);
        user.click(screen.getByRole('button', { name: 'Open section menu' }));
        expect(screen.queryAllByRole('listitem').length).toBe(2);
        expect(screen.getByText('Bar')).toHaveAttribute('to', '/foo/bar');
        expect(screen.getByText('baz')).toHaveAttribute('to', '/foo/baz');
        expect(container.firstChild).toMatchSnapshot();
        user.click(screen.getByRole('button', { name: 'Close section menu' }));
        expect(screen.queryAllByRole('listitem').length).toBe(0);
    });
});
