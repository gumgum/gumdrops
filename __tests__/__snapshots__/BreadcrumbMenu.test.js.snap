// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbMenu renders without option button 1`] = `
<div
  aria-expanded="true"
  aria-haspopup="listbox"
  class="gds-page-header__breadcrumbs-menu"
>
  <button
    aria-label="Close section menu"
    class="gds-page-header__breadcrumbs-menu-dots"
    style="outline: none; padding: 0px;"
    type="button"
  />
  <div
    class="gds-bubble__menu gds-bubble__menu--left gds-bubble__menu--sm gds-bubble__menu--menu-open"
  >
    <ul
      aria-label="Section Menu"
      class="gds-bubble__menu-list"
    >
      <li
        class="gds-bubble__menu-list-item -ellipsis"
      >
        <div
          class="gds-bubble__menu-list-link -text-tr-cap"
          to="/foo/bar"
        >
          Bar
        </div>
      </li>
      <li
        class="gds-bubble__menu-list-item -ellipsis"
      >
        <div
          class="gds-bubble__menu-list-link -text-tr-cap"
          to="/foo/baz"
        >
          baz
        </div>
      </li>
    </ul>
  </div>
</div>
`;
